Adding options in GSI building script (ush/build.sh) for 3DRTMA

GangZhao-NOAA opened the pull request:
<!-- PLEASE READ -->
<!--
Before opening a PR, please note these guidelines:

- Each PR should only address ONE topic and have an associated issue
- No hardcoded or paths to personal directories should be present
- No temporary or backup files should be committed
- Any code that was disabled by being commented out should be removed
-->

**Description**
<!-- Please include relevant motivation and context. -->
<!-- Please include a summary of the change and which issue is fixed. -->
<!-- List any dependencies that are required for this change. -->
**_Motivation_**
GSI, when running in 3DRTMA, needs the support of GSD Cloud Analysis. Because the default configuration in the building script (ush/buiild.sh) does not include the CMake options for GSD Cloud Analysis, therefore when building GSI for 3DRTMA, user has to modify the building script manually to set up the GSD cloud options. We would like to make building GSI more friendly and flexible for various user's configurations. It is better and more flexible by passing the user-specified building options into building script when running the building script, and meanwhile want to avoid modifying the configuration manually in the script for different situation.

**_Method_**
As mentioned in the motivation above, the user-specified building options would be passed into script when running the script. In the first version, the building options are passed as the "positional arguments". Then following the suggestions from code reviewers to make the building script more flexible and friendly, the way of "option arguments" is adopted in building script to pass the user-specified options into script. In the current version, "-c" option is used to get the argument to set up the CMake options, and the arguments could be multiple arguments, which are separated by comma, for example,
` ./build.sh -c c1,c2`
c1 and c2 are two user-specified arguments to set up the special CMake options, which are not set up in default building configurations. 

<!-- Please provide reference to the issue this pull request is addressing. -->
<!-- For e.g. Fixes #IssueNumber -->
Resolves #946

**Type of change**

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)


**How Has This Been Tested?**

<!-- Please describe the tests that you ran to verify your changes and on the platforms these tests were conducted. -->
<!-- Provide instructions so we can reproduce. -->
<!-- Please also list any relevant details for your test configuration -->
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

I built the GSI code with modified building script (ush/build.sh) and run the GSI ctest:
1. CTest with the GSI code built with the default configuration,  that is to run "build.sh" without any arguments just same as before. The code passed the CTest on WCOSS2 Dev machine (Cactus). Here is the ctest running log:
        <details><summary>CTest results on **_WCOSS2_** with GSI code built with default configurations</summary>
        <p>
[gang.zhao@clogin07:build.default] (feature/build_gsi4rtma3d)$ ctest -N
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
  Test # 1: global_4denvar
  Test # 2: rtma
  Test # 3: rrfs_3denvar_rdasens
  Test # 4: hafs_4denvar_glbens
  Test # 5: hafs_3denvar_hybens
  Test # 6: global_enkf
Total Tests: 6
[gang.zhao@clogin04:build] (feature/build_gsi4rtma3d)$ nohup ctest -j 6 > ./ctest_6tasks_run.log 2>&1 &
nohup: ignoring input
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test # 3: rrfs_3denvar_rdasens .............   Passed  798.62 sec
2/6 Test # 2: rtma .............................   Passed  1161.17 sec
3/6 Test # 5: hafs_3denvar_hybens ..............   Passed  1291.43 sec
4/6 Test # 4: hafs_4denvar_glbens ..............   Passed  1470.13 sec
5/6 Test # 6: global_enkf ......................   Passed  1484.70 sec
6/6 Test # 1: global_4denvar ...................   Passed  2048.06 sec
100% tests passed, 0 tests failed out of 6
Total Test time (real) = 2048.10 sec
        </p>
        </details> 
2.  CTest with the GSI code built for 3DRTMA with specific options,  that is to run "_build.sh  **3DRTMA**_". The code also passed the CTest on WCOSS2 Dev machine (Cactus). Here is the ctest running log:
        <details><summary>CTest results on **_WCOSS2_** with GSI code built for _**3DRTMA""_ </summary>
        <p>
[gang.zhao@clogin04:build] (feature/build_gsi4rtma3d)$ git branch
  develop
  feature/build_gsi4rtma3d
  feature/crossvalidation_3drtma
  feature/sfcfull_dumpout
[gang.zhao@clogin04:build] (feature/build_gsi4rtma3d)$ pwd
/lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
[gang.zhao@clogin04:build.default] (feature/build_gsi4rtma3d)$ ctest -N
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
  Test # 1: global_4denvar
  Test # 2: rtma
  Test # 3: rrfs_3denvar_rdasens
  Test # 4: hafs_4denvar_glbens
  Test # 5: hafs_3denvar_hybens
  Test # 6: global_enkf
Total Tests: 6
[gang.zhao@clogin04:build] (feature/build_gsi4rtma3d)$ nohup ctest -j 6 > ./ctest_6tasks_run.log 2>&1 &
[1] 4126990
nohup: ignoring input
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_rdasens
    Start 4: hafs_4denvar_glbens
    Start 5: hafs_3denvar_hybens
    Start 6: global_enkf
1/6 Test # 3: rrfs_3denvar_rdasens .............   Passed  734.02 sec
2/6 Test # 2: rtma .............................   Passed  1102.29 sec
3/6 Test # 5: hafs_3denvar_hybens ..............   Passed  1303.90 sec
4/6 Test # 4: hafs_4denvar_glbens ..............   Passed  1407.09 sec
5/6 Test # 6: global_enkf ......................   Passed  1435.44 sec
6/6 Test # 1: global_4denvar ...................   Passed  2047.09 sec
100% tests passed, 0 tests failed out of 6
Total Test time (real) = 2047.24 sec
        </p>
        </details>

GangZhao-NOAA committed:
modifying ush/build.sh to add options for building GSI for 3D-RTMA.

GangZhao-NOAA committed:
re-adjustment the setup of options to build GSI for 3DRTMA in build.sh

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/build_gsi4rtma3d

GangZhao-NOAA committed:
a trival tweak in build.sh for building gsi for 3drtma

TingLei-NOAA commented on ush/build.sh:
```diff
@@ -2,6 +2,18 @@
 
 set -eux
 
+# checking if the first positional argument is specified to build a special system
```
@GangZhao-NOAA  will it be more flexible /robust to use named arguments than to use positional arguments to pass the parameters?

GangZhao-NOAA replied:
@TingLei-NOAA 
Hi Ting,

Thank you for the suggestion!
You mentioned, 
> use **_named arguments_** than to use positional arguments

Here, the "named arguments", do you mean to build GSI in the way of using the option/flag (for  example, "**_-n_**") to pass the argument, instead of using the position to pass it,  like this? 

"build.sh **_-n 3DRTMA_**" 

I did think about this way. Comparing to the positional arguments, using option is more robust and flexible as you mentioned. It would make the script a little more complicated. Since just a single parameter ("3drtma") is passed into the script, using positional argument is a simple way. If two or more parameters need to be passed (and if need to print out help information to guide the first-time user), I think using the option/flag argument is better. 

Hi, Shun (@ShunLiu-NOAA ) and Russ (@RussTreadon-NOAA), do you suggest me to use positional argument (build.sh 3drtma) or option/flag argument (-n 3drtma) to pass the parameter to building script?

Thanks,
-Gang

ShunLiu-NOAA replied:
@GangZhao-NOAA, I agree with you that the current positional argument should be OK for a single parameter.

RussTreadon-NOAA replied:
My two cents:  Using an  optional command line argument is more robust.  The RDASApp [`build.sh`](https://github.com/NOAA-EMC/RDASApp/blob/develop/build.sh) allows users to modify the build via optional command line arguments.  See [lines 29 - 47](https://github.com/NOAA-EMC/RDASApp/blob/d5c5e69dee514aced1d81b0ef63d68a39d03b929/build.sh#L29-L47) and [71-119](https://github.com/NOAA-EMC/RDASApp/blob/d5c5e69dee514aced1d81b0ef63d68a39d03b929/build.sh#L71-L119) for details.

TingLei-NOAA commented on ush/build.sh:
```diff
@@ -39,6 +51,13 @@ CMAKE_OPTS+=" -DCMAKE_INSTALL_PREFIX=$INSTALL_PREFIX"
 
+# Configure for building GSI with GSD Cloud Analysis
+if [[ -v BUILD_GSI4RTMA3D && ${BUILD_GSI4RTMA3D} =~ [yYtT] ]] ; then
+    BUILD_GSDCLOUD=${BUILD_GSDCLOUD:-"ON"}      # Build GSD Cloud Analysis library
```
Should we directly let BUILD_GSDCLOUD and USE_GSDCLOUD  be "ON"?

GangZhao-NOAA replied:
@TingLei-NOAA 
As far as I know, Global DA (and Hurricane DA maybe?) does not need to build GSI with the support of GSD Cloud Analysis package. Shun told me GSI executable in RRFS-DA also does not include GSD Cloud Analysis. So in current build.sh, the default configurations do not include the setup for GSD Cloud Analysis option. I agree with you that, building GSI with cloud analysis package should not have impact on the result of the Global DA or RRFS DA,. Since it is not necessary, it is better to set it off to avoid any potential issue.
Thanks,
-Gang

TingLei-NOAA replied:
@GangZhao-NOAA  my question is :  I understand, the logic is if BUILD_GSI4RTMA3D = TRUE, the BUILD_GSDCLOUD and USE_GSDCLOUD should be set as "ON" , right?   But use of "S{XX:-"on"} means only if those env variables are not defined yet, they would be set to "ON".   If that is the logic you really want, I am a little confused what the scenario is here.  if it is assumed/required, those variables haven't been defined before this point, {XX:--"ON) is also not needed.  In fact, to me, it is easier to understand to  do a if and else to let them be ON or OFF.

GangZhao-NOAA replied:
@TingLei-NOAA
Good point! I will follow your suggestion to change to **_BUILD_GSDCLOUD="ON"_**  and **_USE_GSDCLOUD="ON"_**.
Thank you, Ting!
-Gang

guoqing-noaa commented:
Sorry that I did not follow all the latest GSI updates but I think GSI should already have the capability to build the GSL cloud analysis. It is needed by RAP/HRRR

I checked the archived comgsi repo and I can see we were able to build cloud analysis using `-DBUILD_GSDCLOUD_ARW=ON`
https://github.com/comgsi/GSI/blob/09f400b8aa3b6dc1df596877ee2e89cf3f690b97/ush/build.comgsi#L154

Something may have changed since then, but I don't think we need to have a RTMA-specific command line option in order to compile cloud analysis. Thanks!

GangZhao-NOAA commented:
> Sorry that I did not follow all the latest GSI updates but I think GSI should already have the capability to build the GSL cloud analysis. It is needed by RAP/HRRR
> 
> I checked the archived comgsi repo and I can see we were able to build cloud analysis using `-DBUILD_GSDCLOUD_ARW=ON` https://github.com/comgsi/GSI/blob/09f400b8aa3b6dc1df596877ee2e89cf3f690b97/ush/build.comgsi#L154
> 
> Something may have changed since then, but I don't think we need to have a RTMA-specific command line option in order to compile cloud analysis. Thanks!

@guoqing-noaa 
Hi Guoqing,

You are right with the building script in the [GSI repository in "comgsi"](https://github.com/comgsi/GSI) community on github. Actually, the initial version of 3DRTMA package was indeed using this "comgsi" when it was on VLab, not on github (back to 2018-2020?). But soon after GSI repository was moved to github, 3DRTMA is using the [GSI repo in NOAA-EMC community](https://github.com/NOAA-EMC/GSI), and in this GSI there is no "build.comgsi" script, just "build.sh". By default, GSD Cloud Analysis options are not set in the build.sh.   I think, RAP/HRRR system uses its own building script (see [line 32](https://github.com/NOAA-EMC/HRRR/blob/v4.1.20/sorc/build_hrrr_gsi.sh#L32)) to build GSI with cloud analysis.
 
Thanks, 
-Gang

guoqing-noaa commented:
@GangZhao-NOAA Thanks for more information.
I would suggest we avoid tightening cloud analysis to 3DRTMA since we can choose NOT to run cloud analysis in 3DRTMA either. It is better to more generic.

If the `-c` command line option is still available,  we can use `build.sh -c` to build the cloud analysis.

Another way is to use an environmental variable directly:
`BUILD_GSDCLOUD=ON build.sh`

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/build_gsi4rtma3d

GangZhao-NOAA committed:
Following reviewers' suggestions (on PR #948), using the option arguments
  to pass the building options into ush/build.sh,
  instead of using positional arguments.  (for issue #946 and PR #948)

RussTreadon-NOAA commented:
**build test**

Install `GangZhao-NOAA:feature/build_gsi4rtma3d` on Cactus.  Execute `./build.sh -h` from `ush/` and get

```
russ.treadon@clogin05:/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr948/ush> ./build.sh -h
++++ readlink -f -n ./build.sh
+++ dirname /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr948/ush/build.sh
++ cd /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr948/ush/..
++ pwd -P
+ readonly DIR_ROOT=/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr948
+ DIR_ROOT=/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr948
+ BUILD_TYPE=Release
+ CMAKE_OPTS=
+ COMPILER=intel
+ BUILD_DIR=/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr948/build
+ INSTALL_PREFIX=/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr948/install
+ GSI_MODE=Regional
+ ENKF_MODE=GFS
+ REGRESSION_TESTS=YES
+ source /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr948/ush/detect_machine.sh
++ [[ -n '' ]]
++ case $(hostname -f) in
+++ hostname -f
++ MACHINE_ID=wcoss2
++ [[ wcoss2 == \U\N\K\N\O\W\N ]]
++ MACHINE_ID=wcoss2
++ [[ wcoss2 != \U\N\K\N\O\W\N ]]
++ return
+ unset OPTS4CMAKE
+ unset SAVELOG
+ unset BUILD_GSDCLDANL
+ getopts c:hlv opt
+ case ${opt} in
+ usage
+ set +x

Usage options: 
  -c additional CMake options                           DEFAULT: <none>; (GSDCLOUD: using GSD Cloud Analysis)
  -v build with verbose output                          DEFAULT: do not use this flag
  -l save the building log files under build directory  DEFAULT: do not use this flag
  -h display this usage/help information and quit
Usage examples: 
  ./build.sh                    # building GSI with default config (for most DA systems, e.g. GDAS, RRFS-DA, etc.)
  ./build.sh -c GSDCLOUD        # building GSI with GSD Cloud Analysis support (only for RAP/HRRR, 3DRTMA, not default)
  ./build.sh -v                 # building GSI with verbose output (not default)
  ./build.sh -l                 # saving the cmake/make/install log into log files under build directory (not default)
```

./build.sh -h` should only return the usage information.  The `-x` option on `set -eux` near the top of `build.sh` is echo all commands to stdout.  This is fine for debugging.  We don't need `-x` in the `build.sh` to be merged into `develop`.

The `-l` option is listed twice.  We only need one occurrence.

Please review `build.sh` and remove unnecessary and/or redundant sections.

RussTreadon-NOAA commented on ush/build.sh:
```diff
@@ -2,10 +2,32 @@
 
+#=================================================================================================#
+usage() {
+#   This usage function is based on the usage function in the build.sh in NOAA-EMC RDASApp 
```
We don't need this comment.  This is the GSI project, not the RDASApp.

RussTreadon-NOAA commented on ush/build.sh:
```diff
@@ -20,6 +42,44 @@ REGRESSION_TESTS=${REGRESSION_TESTS:-"YES"} # Build regression test suite
+shopt -s nocasematch		# enable case-insensitive matching for patterns
+if [[ -v OPTS4CMAKE ]] ; then
+  case "${OPTS4CMAKE}" in
+    gsdcloud|gsdcldanl|gsdcld|gsdcloudanalysis|cloudanalysis|cldanl)
```
Do we want to use all these strings to specify the same build option (BUILD_GSDCLDANL="Yes"?

RussTreadon-NOAA commented on ush/build.sh:
```diff
@@ -20,6 +42,44 @@ REGRESSION_TESTS=${REGRESSION_TESTS:-"YES"} # Build regression test suite
+  esac
+fi
+shopt -u nocasematch		# disable case-insensitive matching for patterns
+unset OPTS4CMAKE
```
We use `OPTS4CMAKE` to set BUILD_GSDCLDANL`.  Why not have the `-c` option directly set this variable instead?  This seems more straightforward.

RussTreadon-NOAA commented on ush/build.sh:
```diff
@@ -47,8 +114,15 @@ CMAKE_OPTS+=" -DGSI_MODE=$GSI_MODE -DENKF_MODE=${ENKF_MODE}"
-make -j ${BUILD_JOBS:-8} VERBOSE=${BUILD_VERBOSE:-}
-make install
+#     specifit options for 3DRTMA
+if [[ -v SAVELOG && ${SAVELOG} =~ [yYtT] ]] ; then
```
Nice feature but a similar result can be achieved with the original `build.sh` via `build.sh > log.build 2>&1`.  `log.build` contains both the build and install information.  `log.build` will be in `/ush`, not `/build`.

GangZhao-NOAA commented:
@RussTreadon-NOAA 
Hi Russ,

Thank you so much for quick review to my updated PR.

The "set -eux" command line have been already in the build.sh script for a while before I made this PR to build.sh script.  
1) If only to avoid the echoed lines when running "build.sh -h" to display the usage info, I can move the section reading the option arguments to the beginning of build.sh script and before the line "set -eux", so when running "build.sh -h", there should be no these annoying echoed command lines.  But there still would be command lines echoed when running "build.sh" in general to build GSI.
2) If you would like to switch on/off the "-x", I can use the "build.sh -v" (for verbose message) to turn on "set -x" in the beginning to echo the command lines , and the default is "set +x" to turn off the echo if not running with "-v".
3) Of just completely remove "set -x" at the beginning of build.sh.

Please let me know which way you would like me to implement.

As for "The -l option is listed twice. We only need one occurrence.". In the display of the usage information, "-l" was listed in the Usage Options session, then listed in the "Usage example" session. So I can remove it from the example session, is that what you mean about the "-l option listed twice"?

Thank you!

-Gang

RussTreadon-NOAA commented:
@GangZhao-NOAA , `build.sh` in `develop` does not have a usage section.  Thus, `set -x` was included to show the user what was happening.  This PR  adds a usage section.  This reduces the need to have the script echo each command.  I suggest replacing `set -eux` with `set -eu`.

Currently, when a user executes `./build.sh -h`, they get
```
Usage options: 
  -c additional CMake options                           DEFAULT: <none>; (GSDCLOUD: using GSD Cloud Analysis)
  -v build with verbose output                          DEFAULT: do not use this flag
  -l save the building log files under build directory  DEFAULT: do not use this flag
  -h display this usage/help information and quit
Usage examples: 
  ./build.sh                    # building GSI with default config (for most DA systems, e.g. GDAS, RRFS-DA, etc.)
  ./build.sh -c GSDCLOUD        # building GSI with GSD Cloud Analysis support (only for RAP/HRRR, 3DRTMA, not default)
  ./build.sh -v                 # building GSI with verbose output (not default)
  ./build.sh -l                 # saving the cmake/make/install log into log files under build directory (not default)
```
In my opinion we do not need the _Usage examples_ section.   The _Usage options_ section is sufficient.  I would remove the _Usage examples_ section.

GangZhao-NOAA committed:
Further modifications in ush/build.sh, based on the suggestions of reviewers
  for PR #948 (and Issue #946)

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/build_gsi4rtma3d

GangZhao-NOAA committed:
A trivial modification in building script (Issue #946, PR #948)

GangZhao-NOAA committed:
another trivial modification to ush/build.sh (Issue #946 and PR #948)

GangZhao-NOAA commented:
@RussTreadon-NOAA @ShunLiu-NOAA @TingLei-NOAA @guoqing-noaa @MatthewMorris-NOAA 
Following your suggestions, I made changes to the building script -- build.sh:
1. using the way of the option argument to pass the parameter into script, instead of the "positional" argument.
    For example, in the first version, to build gsi for 3DRTMA by running "`./build.sh 3DRTMA`". 
    now, to build gsi with GSD cloud analysis support (for 3DRTMA), by running "`./build.sh -c GSDCLOUD`"
2. the "-c" option could pass multiple parameters, which are separated by comma. This makes the building script more flexible and extendable:
    For example,   "`./build.sh -c c1,c2`"   
3. following Russ's suggestion, replacing "set -eux" by "set -eu; set +x" to disable the script debugging (no echo of command lines); and removing the example session in function "usage";

I also merged the latest updates in branch develop of EMC GSI in to the GSI code of this PR #948, and run the 6 regression tests successfully.  Please review the updated PR #948. 
Thank you!

RussTreadon-NOAA commented:
**Alternative solution:  `GSI_MODE`**

[`Russ-Treadon-NOAA/GSI:feature/3drtma`](https://github.com/RussTreadon-NOAA/GSI/tree/feature/3drtma) contains an alternative solution to achieve a 3DRTMA build with GSCCLOUD included.  

GSI `develop` `CMakeLists.txt` currently lists two `GSI_MODE` and four `ENKF_MODE`.  `feature/3drtma` adds a third `GSI_MODE` named `3DRTMA`.  When `GSI_MODE=3DRTMA`, cmake flags `BUILD_GSDCLOUD` and `USE_GSDCLOUD` are set to `ON`.   [Click here](https://github.com/NOAA-EMC/GSI/compare/develop...RussTreadon-NOAA:GSI:feature/3drtma) to see the changes.

I am not requesting that we use the approach in `feature/3drtma`.  I simply wanted to show that there is an alternative solution to issue #946.  This approach was inspired by the global-workflow build of `gsi.x`.  The [global-workflow GSI build script](https://github.com/NOAA-EMC/global-workflow/blob/develop/sorc/build_gsi_enkf.sh) sets [`GSI_MODE=GFS`](https://github.com/NOAA-EMC/global-workflow/blob/9901082002d4bf22a9eca734ebe4d853c4cbad33/sorc/build_gsi_enkf.sh#L30) to get a global-specific GSI build.

GangZhao-NOAA commented:
> **Alternative solution: `GSI_MODE`**
> 
> [`Russ-Treadon-NOAA/GSI:feature/3drtma`](https://github.com/RussTreadon-NOAA/GSI/tree/feature/3drtma) contains an alternative solution to achieve a 3DRTMA build with GSCCLOUD included.
> 
> GSI `develop` `CMakeLists.txt` currently lists two `GSI_MODE` and four `ENKF_MODE`. `feature/3drtma` adds a third `GSI_MODE` named `3DRTMA`. When `GSI_MODE=3DRTMA`, cmake flags `BUILD_GSDCLOUD` and `USE_GSDCLOUD` are set to `ON`. [Click here](https://github.com/NOAA-EMC/GSI/compare/develop...RussTreadon-NOAA:GSI:feature/3drtma) to see the changes.
> 
> I am not requesting that we use the approach in `feature/3drtma`. I simply wanted to show that there is an alternative solution to issue #946. This approach was inspired by the global-workflow build of `gsi.x`. The [global-workflow GSI build script](https://github.com/NOAA-EMC/global-workflow/blob/develop/sorc/build_gsi_enkf.sh) sets [`GSI_MODE=GFS`](https://github.com/NOAA-EMC/global-workflow/blob/9901082002d4bf22a9eca734ebe4d853c4cbad33/sorc/build_gsi_enkf.sh#L30) to get a global-specific GSI build.

@RussTreadon-NOAA 
Hi Russ,

I checked your approach in `feature/3drtma`. That is much better than my current solution. My current modifications would make the build.sh script messy and unstructured. I think my current PR #948 might not be worth to merge into GSI. I am going to close this PR #948. 

Thank you very much for proposing the alternative approach!
